docs/concepts.md

```diff
@@ -0,0 +1,123 @@
+# Concepts
+
+PrairieLearn is extremely flexible and powerful, but this flexibility can be overwhelming at first. This document provides an overview of the key concepts in PrairieLearn, and how they fit together.
+
+## High-level overview
+
+A [**Course**](./course.md) in PrairieLearn is composed of **Course Instances**, **Questions**, and **Custom Elements**. Each [course instance](./courseInstance.md) is an offering of a course in a particular semester/term. For example, a course "CS 225" might have instances "CS 225 Spring 2022" and "CS 225 Fall 2022". A course instance contains all the **Assessments** for that particular term. The course itself contains the questions and custom elements that are shared across all instances. You can think of the course as a bank of questions and elements, that are then used to create assessments for each instance of the course.
+
+=== "Concept Map"
+
+    ```d2
+    --8<-- "docs/diagrams/concept-map.d2"
+    ```
+
+=== "Example for the `PrairieLearn 101` course"
+
+    ```d2
+    --8<-- "docs/diagrams/concept-map-example.d2"
+    ```
+
+### Questions
+
+[**Questions**](./question.md) are all independent, and can be graded [using Python](./question.md), or using an [external code autograder](./externalGrading.md). They are written in HTML (as a [Mustache template](https://mustache.github.io/mustache.5.html)) and generated / graded using Python.
+
+=== "Example `question.html`"
+
+    ```html title="question.html"
+    <pl-question-panel>
+        If $x = {{params.x}}$ and $y$ is {{params.operation}} $x$, what is $y$?
+    </pl-question-panel>
+
+    <pl-number-input answers-name="y" label="$y =$"></pl-number-input>
+    <pl-submission-panel> {{feedback.y}} </pl-submission-panel>
+    ```
+
+=== "Example `server.py`"
+
+    ```python title="server.py"
+    def generate(data):
+        data["params"]["x"] = random.randint(5, 10)
+        data["params"]["operation"] = random.choice(["double", "triple"])
+
+        if data["params"]["operation"] == "double":
+            data["correct_answers"]["y"] = 2 * data["params"]["x"]
+        else:
+            data["correct_answers"]["y"] = 3 * data["params"]["x"]
+
+    def parse(data):
+        if "y" not in data["format_errors"] and data["submitted_answers"]["y"] < 0:
+            data["format_errors"]["y"] = "Negative numbers are not allowed"
+
+    def grade(data):
+        if math.isclose(data["score"], 0.0) and data["submitted_answers"]["y"] > data["params"]["x"]:
+            data["partial_scores"]["y"]["score"] = 0.5
+            data["score"] = 0.5
+            data["feedback"]["y"] = "Your value for $y$ is larger than $x$, but incorrect."
```

Comment:
This isn't really representative of what instructors would usually put in questions; I'd be inclined to remove this and the corresponding `<pl-submission-panel>` from `question.html`.